### ISSUE TITLE
Sparse checkout support

### ISSUE THREAD

**Issue by jspahrsummers** [libgit2/libgit2#2263]
A user can set `core.sparsecheckout` to `true`, then populate `.git/info/sparse-checkout` with the specific subtrees that should be checked out. See http://jasonkarns.com/blog/subdirectory-checkouts-with-git-sparse-checkout/ for a basic explanation.

I don't know whether it makes sense for libgit2 to support this in its own checkout operations, but sparse checkouts should be taken into consideration for diffs and status (in that omitted subtrees should not appear).

**Comment by calavera** [comment]
I understand for the lack of traffic here that this is not in any roadmap, but I thought I could ask again. @ethomson, @carlosmn is there any possibility that this can be implemented? I don't really understand the whole scope of this request, so "no" is a very valid answer.

**Comment by carlosmn** [comment]
It is highly unlikely that either of us will work on this. Maybe @johnhaley81 is interested in this since he works on a client-side app.

**Comment by johnhaley81** [comment]
It won't be anytime soon but that is something that I'm looking to tackle eventually :)

**Comment by jlindner** [comment]
Is this any update on this functionality possibly making it into libgit2?  This has pretty much become a show stopper for us to use git since we have massive mono repos, but only need to perform diffs of certain directories.  Currently, this lack of support has kept us from moving everything from Subversion to git.  Since this was talked about a couple of years ago, I was hoping somebody would maybe have an update on this.  

Thanks

**Comment by ethomson** [comment]
@tiennou has done some work here, but this is an insanely complex topic and remains unimplemented.

**Comment by tiennou** [comment]
I think you're mismatching with shallow support ? At least I don't remember going anywhere deep down checkout yet 😉.

From a cursory look at it, sparse checkouts ought to be easier to implement than shallow is (there seem to be no transport-level/pkt-protocol changes needed, as that the repo is still downloaded in full).

I wonder though, doesn't diff provide a pathspec limiter already ? Am I missing something for your use case @jlindner ?

**Comment by ethomson** [comment]
Yes I was conflating the two; I read this issue too quickly and thought it was about shallow not sparse.  Sorry!

**Comment by jlindner** [comment]
Yeah.  I definetly need the sparseCheckout ability I believe.  Basically, I have about 1500 directories(each directory contains a c++ project) in one git repo.  I have a need to extract one directory(one project), without the rest of the directories(projects), and be able to get data from it.  I know I can use the git command line to get to this data, but I was hoping that I could do it from the libgit2 functionality, without shelling out to the command line.

**Comment by jlindner** [comment]
With the diff, which I haven't looked into that much, will it allow me to extract from the git repo, a specific directory, and a specific commit?  I am not needing to diff, but if that has the ability to get that out where I could access the files, that might work.  

Basically, what is happening is for a directory/commit, I extract the files in that location, and run some scanners on it making sure they don't contain security flaws and a few other checks.  Since the user gives me the directory and commit, for performance reasons, I would like to just get that out, and nothing else.  Cloning the whole git repository locally isn't great either, as it is pretty good sized.

**Comment by tiennou** [comment]
Sorry, I'm still unclear on what you need, and your clarifications seems to imply you *really* need the files to exist in the filesystem at some point… If that's the case, that's sparse checkout. If that's not, you might be able to get to the "contents" of the directory/files by looking up the commit, traverse its tree to get at the "project" path, and you'll have another tree whose contents are accessible (in memory at least, if your scanner supports that).

Another caveat is that all this will require a git repository "locally" (though what you meant eludes me), ie. libgit2 *will* need to open a "clone" of your repository, either the canonical one (because your application is "local" to the machine that has the repository), or a brand new one (ie. clone it).

**Comment by jlindner** [comment]
Do you know a better way to accomplish what I am needing to do?  Do you think the sparseCheckout is what I need, or do you know of a better way to basically export one directory from a larger github repository.  I am open to any ideas as to the best way to get the files out.  I don't need a checkout/clone, I just need the actual files that I have checked in.

**Comment by jlindner** [comment]
Using the sparsecheckout from git.exe, I have been able to get just one directory out, the size of the entire repo is about 6GB currently, with several thousand projects.  The local repo created for the sparsecheckout to work, ends up being about 115k, so it is not bringing down the entire repo, it just brings down that one directory, checked out.  It does create the .git folders, which I can live with, because I can just delete them, but it isn't bringing down the entire repo, and taking up 6GB of disk space.  I was hoping for an API that would be able to perform this action, since I am actually doing this task from a rest web service that the user calls to perform the action.  Shelling out the command line in a web service may not be the worlds smartest idea, but if there is no other way around it, that may be my only option since none of the APIs seem to support this task.

**Comment by ethomson** [comment]
> It does create the .git folders, which I can live with, because I can just delete them, but it isn't bringing down the entire repo

That _is_ the repo.

I’m confused.  What is the size of the `.git` directory with git?  And with libgit2?

**Comment by tiennou** [comment]
Can you provide the git commands you're executing ? That would clarify *a lot*, as right now unless you're running those commands on the same machine the repo is in (triggering an "optimized" local clone), you're likely going through the git protocol, which has no support for fetching only some specific paths.

**Comment by jlindner** [comment]
After doing more testing with the sparseCheckout, I have figured out that it isn't going to work for what I need.  I was testing the sparseCheckout on a little repository, with only a few directories in it.  It seemed really fast, and I wasn't looking at the logs of the action close enough, to realize that is was in fact cloning the entire repo locally, before it ran the sparseCheckout functionality to give me just the files I was wanting.  As soon as I pointed it at the real repository, you are correct, it clones the whole thing locally first, and then sets the directory to only what I wanted.  Sorry for the confusion. I don't think this is going to work for what I was needing, and we are going to need to take a different approach on the repo.

Thanks for all your time.  Keep up the good work.

James

**Comment by ethomson** [comment]
Thanks for the confirmation, @jlindner.  What you've experienced aligns with my expectations.

[Work on "narrow" or "partial" clone support in git is continuing](https://github.com/git/git/blob/master/Documentation/technical/partial-clone.txt) but this is in-progress and not something that libgit2 supports yet.  I would encourage you to check out the appropriate API for your hosting provider in the meantime.

Thanks!

**Comment by mathomp4** [comment]
Hello all. The linked issue above is from me. The project I'm working on uses sparse checkout and apparently powerlevel10k prompt doesn't handle sparse checkout because libgit2 doesn't handle it.

**Comment by mathomp4** [comment]
As an aside, it looks like git itself has been focusing on sparse checkout:

https://github.blog/2020-01-17-bring-your-monorepo-down-to-size-with-sparse-checkout/
https://github.blog/2020-03-22-highlights-from-git-2-26/#updates-to-git-sparse-checkout

So hopefully that will make it more common and might get sparse-checkout support in libgit2 up the priority tree.

**Comment by briancullinan** [comment]
Is it possible to use 
`git pull --depth=1 origin master`
with 
`git clone --filter=blob:none --sparse git@gitlab.com:gitlab-com/www-gitlab-com.git`
As a workaround for sparse checkout?

https://docs.gitlab.com/ee/topics/git/partial_clone.html
https://stackoverflow.com/questions/600079/git-how-do-i-clone-a-subdirectory-only-of-a-git-repository/28039894#28039894

**Comment by jairbubbles** [comment]
Looks like there is a PR on the subject: #5833

**Comment by zentron** [comment]
PR https://github.com/libgit2/libgit2/pull/5833 and https://github.com/libgit2/libgit2/pull/6169 both look for provide a workable solution to sparse checkouts. Is there anything further waiting for these to progress and merge?

**Comment by ethomson** [comment]
@Nikitae57 have you tested them? Do they work? What sort of problems do they still have? Where are the performance issues?

**Comment by mathomp4** [comment]
I am thinking that #6394 by @YuKitsune is probably the PR to focus on. I think it's a successor to #5833 by @jochenhz?

**Comment by wtfacoconut** [comment]
Raising my hand to keep this issue alive. Sparse checkout is indeed a much needed capability.

**Comment by MadL1me** [comment]
Extremely relevant feature for me as well. I'm required to use plain git in cli instead of libgit2, because it does not supports with currently.
